EgressIP downstream tests for OVN Kubernetes

andreaskaris opened the pull request:
Implementation of the EgressIP downstream tests for OVN Kubernetes.
Given that these tests manipulate node and hostSubnet annotations, they must run without competition 
as part of the serial conformance suite.
Test behavior can be changed using several env variables:
EGRESSIP_TARGET_PROTOCOL, EGRESSIP_TARGET_HOST, EGRESSIP_TARGET_PORT

This is a first take on the tests and only tests IPv4 for OVNKubernetes and OpenShiftSDN on Azure, GCP and AWS. 

**Note:** The `vendor-update` label here is a legacy label that the bot added for an earlier version of this PR. The newest version does not modify the vendor folder, yet the bot still reports this.

**Note:** These tests only run on AWS, Azure and GCP and only run in serial. They skip anything baremetal and thus failing baremetal tests would not be due to these tests.
See: https://github.com/openshift/origin/pull/26999/commits/c69664a6d037b433a22036bdce9d41b5829d28bc#diff-dfcd246ff964649f05970d223aea24d306b7c748c8c9c094455fd01c34411303R109

andreaskaris commented:
Azure:
===============================================

Test: ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets pods should have the assigned EgressIPs and EgressIPs can be deleted and recreated [Disabled:Unsupported]"
Result:  The test works (see notes), but Azure has issues with cloudprivateipconfigs deletion https://bugzilla.redhat.com/show_bug.cgi?id=2073045

Test: ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets pods should keep the assigned EgressIPs when being rescheduled to another node [Disabled:Unsupported]"
Result: PASS

Test: ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets pods should have the assigned EgressIPs and EgressIP objects can be updated [Disabled:Unsupported]"
Result: PASS

Test: ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets only pods matched by the pod selector should have the EgressIPs [Disabled:Unsupported]"
Result: PASS

GCP:
===============================================

Test: ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets pods should have the assigned EgressIPs and EgressIPs can be deleted and recreated [Disabled:Unsupported]"
Result: PASS

Test: ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets pods should keep the assigned EgressIPs when being rescheduled to another node [Disabled:Unsupported]"
Result: PASS

Test: ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets pods should have the assigned EgressIPs and EgressIP objects can be updated [Disabled:Unsupported]"
Result:  PASS

Test: ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets only pods matched by the pod selector should have the EgressIPs [Disabled:Unsupported]"
Result: PASS

AWS (tests should be TCP to own route):
===============================================

Test: ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets pods should have the assigned EgressIPs and EgressIPs can be deleted and recreated [Disabled:Unsupported]"
Result: PASS

Test: ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets pods should keep the assigned EgressIPs when being rescheduled to another node [Disabled:Unsupported]"
Result: PASS 

Test: ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets pods should have the assigned EgressIPs and EgressIP objects can be updated [Disabled:Unsupported]"
Result: PASS

Test: ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets only pods matched by the pod selector should have the EgressIPs [Disabled:Unsupported]"
Result: PASS

tssurya commented:
/cc

andreaskaris commented:
CC @jluhrsen @flavio-fernandes @kyrtapz

jluhrsen commented on test/extended/networking/util.go:
```diff
@@ -417,3 +419,120 @@ func InOVNKubernetesContext(body func()) {
 }
+
+// SubnetIPs enumerates all subnets in an ip net (starting with the provided IP address).
+func SubnetIPs(ipnet net.IPNet) ([]net.IP, error) {
```
code, to me, seems fine. first thing I wondered was if these functions aren't already available somewhere/somehow
in some other utility or library. but, assuming you already checked and nothing obvious stood out.

andreaskaris replied:
I already checked and couldn't find anything (which doesn't mean that it's not there). I'll remove code duplication for DetermineImageFromRelease, though.

jluhrsen commented on test/extended/networking/egressip.go:
```diff
@@ -0,0 +1,654 @@
+
+		workerNodesOrdered    []corev1.Node
+		workerNodesOrderedStr []string
+		// egressIPNodesOrdered    []corev1.Node
```
any reason to keep these commented lines? some below too

andreaskaris replied:
Removed.

jluhrsen commented on test/extended/networking/egressip.go:
```diff
@@ -0,0 +1,654 @@
+			g.By("Determining the target protocol, host and port")
+			targetProtocol, targetHost, targetPort, err = getTargetProtocolHostPort(oc, hasIPv4, hasIPv6, cloudType)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			framework.Logf("Testing against: Protocol %s, TargetHost: %s, TargetPort: %d", targetProtocol, targetHost, targetPort)
```
add cloudType in this Logf?

jluhrsen commented on test/extended/networking/egressip.go:
```diff
@@ -0,0 +1,654 @@
+			workerNodesOrdered, err = getWorkerNodesOrdered(clientset)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			for _, s := range workerNodesOrdered {
+				workerNodesOrderedStr = append(workerNodesOrderedStr, s.Name)
```
workerNodesOrderedNames instead of ...Str?

andreaskaris replied:
Done

jluhrsen commented on test/extended/networking/egressip.go:
```diff
@@ -0,0 +1,654 @@
+			}
+
+			g.By("Selecting the EgressIP nodes and a non-EgressIP node")
+			// nonEgressIPNode = workerNodesOrdered[0]
```
rm comments

andreaskaris replied:
Done

jluhrsen commented on test/extended/networking/egressip.go:
```diff
@@ -0,0 +1,654 @@
+				var targetPort int
+
+				g.By("Selecting a single EgressIP node, and one node per source deployment")
+				// requires a total of 3 worker nodes
```
are you only checking > 1, or do you want to explicitly check for 3

andreaskaris replied:
I just need more than 1 target node (the one hosting the host networked pod) and 2 nodes for the egress ip tests --> a total of 3 worker nodes. I updated the comment in the new version, that'll hopefully be easier to understand.

jluhrsen commented on test/extended/networking/egressip.go:
```diff
@@ -0,0 +1,654 @@
+					framework.Logf("clientIPSet is: %v", clientIPSet)
+					o.Expect(len(clientIPSet)).Should(o.BeNumerically(">", 0))
+					// Note: it's a bug if we see an egressIP here:
+					// https://bugzilla.redhat.com/show_bug.cgi?id=2070929
```
this bug is still marked as assigned (not fixed), so does this fail every time? I saw in your comments that most of these tests are PASS.

andreaskaris replied:
So to begin with, I will disable those internal tests because they fail until 2070929 is fixed. As soon as Surya fixes that bug, I will talk to her and Tim what to do with this here (I updated the comment for clarification):
~~~
269                                         // Note: my interpretation is that it's a bug if we see an egressIP here:
270                                         // We should never see egressIPs when querying internal targets:
271                                         // https://bugzilla.redhat.com/show_bug.cgi?id=2070929
272                                         // However, this was still a subject of discussion. When we enable these tests after
273                                         // we fix 2070929, decide if we want to see EgressIPs here or not and possibly remove
274                                         // this verification.
275                                         g.By("Making sure that EgressIPs were not part of the response")
276                                         framework.Logf("egressIPSet is: %v", egressIPSet)
277                                         framework.Logf("clientIPSet is: %v", clientIPSet)
278                                         o.Expect(len(clientIPSet)).Should(o.BeNumerically(">", 0))
279                                         o.Expect(AntiDeepEqual(egressIPSet, clientIPSet)).To(o.BeTrue())
~~~

I believe this is a bug, but when we talked about this, the others seemed to disagree. So there's a chance that I'll remove the above part, but for as long as the internal tests are disabled, I prefer to keep that part of the code and defer the final decision to a later point in time.

jluhrsen commented on test/extended/networking/egressip.go:
```diff
@@ -0,0 +1,654 @@
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				// Run this twice to make sure that repeated EgressIP creation and deletion works.
+				iterations := 5
```
what is "iterations"? is it num packets sent for the packet sniffer work?

andreaskaris replied:
I renamed this to: `numberOfRequestsToSend`

jluhrsen commented on test/extended/networking/egressip.go:
```diff
@@ -0,0 +1,654 @@
+				err = ioutil.WriteFile(tmpDirEgressIP+"/"+egressIPYaml, []byte(egressIPConfig), 0644)
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				// Run this twice to make sure that repeated EgressIP creation and deletion works.
```
don't see this happening twice. copy/paste from above?

andreaskaris replied:
yep, removed

jluhrsen commented on test/extended/networking/egressip.go:
```diff
@@ -0,0 +1,654 @@
+		return true
+	}, time.Duration(timeout)*time.Second, 5*time.Second).Should(o.BeTrue())
+
+	// There's some additional delay in play here. If we probe right after the above check,
```
don't like having a sleep at the end of a function. would it make sense to wrap whatever the "probe" is in some
Eventually() that checks once a second for 15s? or something along those lines.

andreaskaris replied:
Done (in both occurrences where I used a sleep instead of a retry mechanism)

jluhrsen commented on test/extended/networking/egressip_helpers.go:
```diff
@@ -0,0 +1,1330 @@
+
+const (
+	nodeEgressIPConfigAnnotationKey = "cloud.network.openshift.io/egress-ipconfig"
+	// egressIPAgnImage                = "k8s.gcr.io/e2e-test-images/agnhost:2.36" // TODO: remove
```
remove these commented lines?

andreaskaris replied:
Done

jluhrsen commented on test/extended/networking/egressip_helpers.go:
```diff
@@ -0,0 +1,1330 @@
+
+const (
+	tcpCaptureScript = `#!/bin/bash
+tcpdump -nne -i any -l -s 0  'port %d and tcp[((tcp[12:1] & 0xf0) >> 2):4] = 0x47455420' | awk '{print $9 " " $(NF-1)}'
```
I know it would require a bigger refactor, but since it popped in to my head I figured to note it. I remember writing tests using tcpdump and using the -w <filename> option (as a background process), then doing a 'killall tcpdump' to finish it when it should be done writing. then tcpdump -r <filename> to read the packets back and parse what I needed.
That helped with some of the weirdness I would see when not actually capturing packets it should have.

BUT, looks like you have some good methods here as well.

jluhrsen commented:
in your earlier [comment](https://github.com/openshift/origin/pull/26999#issuecomment-1093053975) you note that most tests PASS. It is not reliable enough? just not fully understanding why we
want to have these skipped by default initially?  And if so, I didn't quite catch the plan to have these run some other
way, other than manually?

andreaskaris commented:
/test ?

andreaskaris commented:
/meow caturday

andreaskaris commented:
/test ?

andreaskaris commented:
@jluhrsen I think this is ready for another review, I addressed your concerns, added 3 tests for OpenShift SDN, and tested against the aforementioned cloud platforms and network plugins. Baremetal is currently skipped on purpose as the tests require the cloudprivateipconfig CRD which is only available on cloud platforms (that also means that the tests are only ipv4).

jluhrsen commented:
> > in your earlier [comment](https://github.com/openshift/origin/pull/26999#issuecomment-1093053975) you note that most tests PASS. It is not reliable enough? just not fully understanding why we want to have these skipped by default initially? And if so, I didn't quite catch the plan to have these run some other way, other than manually?
> 
> So I had thought about setting the tests as disabled. Then, adding new test cases in github.com/openshift/release where the tests would be triggered individually in a bash script by their name, e.g.:
> 
> ```
> ./openshift-tests run-test "[sig-network][Feature:EgressIP] when using openshift ovn-kubernetes when testing to external targets pods should keep the assigned EgressIPs when being rescheduled to another node [Disabled:Unsupported]"
> ```
> 

I don't think this is a normal (or if it's anywhere) way of running tests in the release repo. I mean we could create
a whole new job that would do it, but that's a lot of overhead (new cluster, new job to maintain, etc).



> That test case in openshift/release could then be triggered by default or manually and could be marked as optional.
> 
> There are currently 2 problems that I see: a) The tests deal with node annotation and cannot run in parallel ; so the safest way of running them is [Serial]. That adds a lot of time to all unit tests, as each test takes between 1:30 and a bit more than 2 minutes to complete (that's on AWS, on Azure, I expect this to be slower) b) I'm a bit afraid that we enable these tests, and then we cause CI failure for other projects, and then we might have to roll back.

I think we should make a plan to get these un-excluded and running as part of the serial jobs. 6 new tests right? That
is 15-20m longer. If that's too long, maybe we can find a way to optimize and  shorten them, but for now we can
see if it's ok.

I think we can try to get this PR merged and then lay out a plan to get these running automatically in all
the serial suites. If there is some case (e.g., certain platform, or project) where these can't pass for some
reason, we can tweak those jobs to exclude specific tests.

Let's make some jira stories to figure this out so we don't lose track:
- run these tests multiple times on each of AWS, GCP, Azure to make sure they pass reliably
- figure out which projects have the serial suite running in their presubmits and see if we have any failures there to worry about.

jluhrsen commented:
/lgtm

andreaskaris commented:
/assign @sjenning

stbenjam commented:
/remove-label vendor-update

andreaskaris commented:
/retest-required

jluhrsen commented:
/lgtm

jluhrsen commented:
@deads2k, do we need a bz for this, or can we override that bugzilla/valid-bug label requirement?
The work was tracked in jira [here](https://issues.redhat.com/projects/SDN/issues/SDN-1709?filter=allopenissues)

deads2k commented:
bypassing bug requirement to add tests
